Convert HTML to Markdown (PM_COLLECT_PROC)

ratijas opened the pull request:
Fixes a lot of formatting wrongdoings, as well as renames
and reorders parameters consistently with callback signature.

Regarding this parameters:

    _In_opt_ LPWSTR lpValueName,
    ...
    _Inout_ LPDWORD lpNumObjectTypes

The query string parameter *lpValueName* is annotated as optional
(nullable), however I have not found a signle proof of that in either
documentation or in code. Queries against performance counters HKEYs
via RegQueryValueExW are meaningless without *lpValueName*, which is
exactly our query string. Running such null query results in
`Error Code 0x00000057: The parameter is incorrect.`

Number of objects written, *lpNumObjectTypes*, is specified as in/out.
I believe, this was an oversight. Documentation does not explain how
to use it as an input anyway.

ratijas committed:
Convert HTML to Markdown (PM_COLLECT_PROC)

Fixed a lot of formatting wrongdoings, as well as renamed
and reordered parameters consistently with callback signature.

Two parameters, namely *lpValueName* and *lpNumObjectTypes* were
annotated incorrectly in source code (or at least the way they were
does not make sense), so a [!NOTE] were added to each one to mitigate
that oversight.

PRMerger16 commented:
@ratijas : Thanks for your contribution! The author(s) have been notified to review your proposed change.

ratijas commented:
I'd wish there were Github issues available, but here we go.

We need to port <a href="/previous-versions/windows/desktop/legacy/aa372200(v=vs.85)">OpenPerformanceData</a> page to new documentation system. It is the last remaining function there, and markdown sucks at links with parentheses.

Karl-Bridge-Microsoft commented:
Thanks for the updates! However, we should reflect the source definition in the documentation. The qualifications you identify could be added as notes. Here's the definition again:

typedef DWORD (APIENTRY PM_COLLECT_PROC) (_In_opt_ LPWSTR lpValueName,
                                          _Inout_
                                            _At_(*lppData, _Pre_writable_byte_size_(*lpcbTotalBytes) _Post_writable_byte_size_(*lpcbTotalBytes))
                                            _At_(_Old_(*lppData), _Post_writable_byte_size_(*lpcbTotalBytes))
                                                  LPVOID *lppData,
                                          _Inout_ LPDWORD lpcbTotalBytes,
                                          _Inout_ LPDWORD lpNumObjectTypes);

ratijas commented:
> Here's the definition again:

It is actually with those ugly underscore \_prefixes\_and\_suffixes\_ everywhere, but yeah, this is it.

I think this is a perfect example why documentation should be generated _strictly_ from the source code (~cough cough rust~). Also, I don't understand where that #Syntax block magically comes from.

So, what is the best course of amendments? Reflect actual incorrect "attributes" as they are, and add my slightly modified commit message as a `> [!NOTE]  AcTuAlLy, ...` ?

Karl-Bridge-Microsoft commented:
Yes, that's what I tried to convey in my previous response. The attributes should reflect reality, while any note/qualification should address a concrete concern/error condition while being concise and to the point.

ratijas commented:
Sorry. Just wanted to make it super clear for myself, so I won't end up doing needless work.

PRMerger15 commented:
@ratijas : Thanks for your contribution! The author(s) have been notified to review your proposed change.

ratijas commented:
Reworked, amended. Check it out now.

BTW, I still stand by [this statement](https://github.com/MicrosoftDocs/sdk-api/pull/476#issuecomment-698659228):
> We need to port OpenPerformanceData page to new documentation system. It is the last remaining function there, and markdown sucks at links with parentheses.

Are there sources of old documentation available? Can I get help porting page metadata etc.?

ratijas commented:
@Karl-Bridge-Microsoft Hi, How is it going?

Karl-Bridge-Microsoft commented:
Hi @ratijas, thanks for your efforts with this topic. Your updates are extensive and require significant review and refinement. Some definitely cannot be published as is. I'll address as time permits.

Karl-Bridge-Microsoft committed:
Update nc-winperf-pm_collect_proc.md

Some initial edits.

Karl-Bridge-Microsoft commented on sdk-api-src/content/winperf/nc-winperf-pm_collect_proc.md:
```diff
@@ -47,154 +47,102 @@ api_name:
+Consumer-allocated buffer that contains the performance data.
 
-On input, specifies the size, in bytes, of the <i>pData</i> buffer. 
+Here, *pData* refers to the pointer pointed by *lppData* (`*lppData = pData`).
```
Where is pData? First mention and no reference.

ratijas replied:
Now that I think about it, it'd better be paraphrased like this:

> Let *pData* be the pointer pointed by *lppData*.

My intention was to lift one layer of confusing indirection from the reader.

ratijas commented on sdk-api-src/content/winperf/nc-winperf-pm_collect_proc.md:
```diff
@@ -143,9 +140,9 @@ The following tests are performed only if test level 1 is used:
 ## -see-also
 
-- <a href="/previous-versions/windows/desktop/legacy/aa372200(v=vs.85)">OpenPerformanceData</a>
+- [OpenPerformanceData](/previous-versions/windows/desktop/legacy/aa372200)
```
No. This is wrong, 404. The `(v=vs.85)` part is required.

I intentionally left it as an `<a>` tag because parentheses aren't friendly with markdown syntax. Then again, we need to port this last remaining article to this repository.

Karl-Bridge-Microsoft replied:
ok, reverted.

ratijas commented on sdk-api-src/content/winperf/nc-winperf-pm_collect_proc.md:
```diff
@@ -59,23 +59,20 @@ The **CollectPerformanceData** function is a placeholder for the application-def
 Type: **LPWSTR**
 
-Null-terminated string that contains the query string (for example, "Global" or "238") passed to the [RegQueryValueEx](/windows/desktop/api/winreg/nf-winreg-regqueryvalueexa) function. For the query string format, see [Using the Registry Functions to Consume Counter Data](/windows/desktop/PerfCtrs/using-the-registry-functions-to-consume-counter-data).
+Null-terminated string that contains the query string (for example, "Global" or "238") passed to the [RegQueryValueEx](/windows/desktop/api/winreg/nf-winreg-regqueryvalueexa) function. For a list of possible values for *lpValueName*, see [Using the Registry Functions to Consume Counter Data](/windows/desktop/PerfCtrs/using-the-registry-functions-to-consume-counter-data).
```
I'm not implying that one is better that the other, but there are places in the documentation where I copied this phrase from. For better consistency we should either leave it as it is, or change elsewhere as well, I think.

Karl-Bridge-Microsoft replied:
We can address updates to the other topics in a separate PR.

ratijas commented on sdk-api-src/content/winperf/nc-winperf-pm_collect_proc.md:
```diff
@@ -59,23 +59,20 @@ The **CollectPerformanceData** function is a placeholder for the application-def
 Type: **LPVOID \***
 
-Consumer-allocated buffer that will contain the performance data.
+Consumer-allocated buffer that contains the performance data.
```
Big oof. Sorry, my bad.

ratijas reviewed (commented):
Hi, Karl! Sorry for the delay. Here goes my review.

ratijas commented:
Hi, @Karl-Bridge-Microsoft, it's been a while. I'm starting to worry.

What do you think of my comments?

Karl-Bridge-Microsoft committed:
Update nc-winperf-pm_collect_proc.md

link revised

Karl-Bridge-Microsoft committed:
Update nc-winperf-pm_collect_proc.md

ugh. link re-revised.

PRMerger18 commented:
@Karl-Bridge-Microsoft : Thanks for your contribution! The author(s) have been notified to review your proposed change.

Karl-Bridge-Microsoft commented:
Hi @ratijas - as I remarked, the updates are quite extensive and need thorough review. Thanks for your patience.

ratijas commented:
Hi @Karl-Bridge-Microsoft,

It's OK, no worries. Take your time.

Do you think you could provide me with a source of OpenPerformanceData page? While you are busy with other things, I'm willing to put some effort in porting it here. That way, we could finally replace that ugly old link for good.  But I can't do it completely on my own, as I am missing yaml metadata which cannot be retrieved from a rendered webpage.

> link revised
> ugh. link re-revised.

It's gonna be such squash & very merge :rofl: wow

> the updates are quite extensive and need thorough review.

I'd be happy to help in any way I can. If you need my cooperation, don't hesitate to ping me — it's not like I'm busy with anything urgent recently.

Best,
Ivan

Karl-Bridge-Microsoft committed:
Update nc-winperf-pm_collect_proc.md

More revisions for conciseness, grammar, and consistency within topic. I think this is the last.

ratijas commented on sdk-api-src/content/winperf/nc-winperf-pm_collect_proc.md:
```diff
@@ -90,29 +88,29 @@ If this function fails, set *lpcbTotalBytes* to zero.
 ## -remarks
 
-If the requested objects specified in the *lpValueName* parameter do not correspond to any of the object indexes that your performance DLL supports, leave the *pData* parameter unchanged, and set the *lpcbTotalBytes* and *lpNumObjectTypes* parameters to zero. This indicates that no data was returned.
+If the requested objects specified in the *lpValueName* parameter do not correspond to any of the object indexes that your performance DLL supports, leave the *pData* parameter (where *pData* refers to the pointer pointed to by *lppData*) unchanged, and set the *lpcbTotalBytes* and *lpNumObjectTypes* parameters to zero. This indicates that no data was returned.
```
I feel uneasy about splitting the sub-sentence like that. May I suggest the following?
 
```suggestion
If the requested objects specified in the *lpValueName* parameter do not correspond to any of the object indexes that your performance DLL supports, leave the *pData* parameter unchanged (where *pData* refers to the pointer pointed to by *lppData*), and set the *lpcbTotalBytes* and *lpNumObjectTypes* parameters to zero. This indicates that no data was returned.
```

ratijas reviewed (commented):
LGTM. Just one nit below, otherwise I don't see any semantic issues.

Karl-Bridge-Microsoft committed:
Update sdk-api-src/content/winperf/nc-winperf-pm_collect_proc.md

Co-authored-by: ratijas <me@ratijas.tk>

PRMerger15 commented:
@Karl-Bridge-Microsoft : Thanks for your contribution! The author(s) have been notified to review your proposed change.
